src/TestCase.ts

```diff
@@ -0,0 +1,241 @@
+  };
+}
+
+export function serializePosition(position: Position): SerializedPosition {
+  return { line: position.line, character: position.character };
+}
+
+type TestCaseCommand = {
+  actionName: ActionType;
+  partialTargets: PartialTarget[];
+  extraArgs: any[];
+};
+
+type TestCaseContext = {
+  talonCommand: string;
+  thatMark: ThatMark;
+  targets: Target[];
+  navigationMap: NavigationMap;
+};
+
+type TestCaseSnapshot = {
+  document: string;
+  clipboard: string;
+  visibleRanges: SerializedRange[];
+  selections: SerializedSelection[];
+  thatMark: SerializedSelection[] | null;
+};
+
+type DecorationRanges = { [coloredSymbol: string]: SerializedRange };
+
+export type TestCaseFixture = {
+  talonCommand: string;
+  command: TestCaseCommand;
+  targets: Target[];
+  languageId: string;
+  marks: DecorationRanges;
+  initialState: TestCaseSnapshot;
+  finalState: TestCaseSnapshot;
+  returnValue: any;
+};
+
+export default class TestCase {
+  talonCommand: string;
+  command: TestCaseCommand;
+  languageId: string;
+  targets: Target[];
+  marks: DecorationRanges;
+  context: TestCaseContext;
+  initialState: TestCaseSnapshot | null = null;
+  finalState: TestCaseSnapshot | null = null;
+  returnValue: any = null;
+
+  constructor(command: TestCaseCommand, context: TestCaseContext) {
+    const activeEditor = vscode.window.activeTextEditor!;
+    const { navigationMap, targets, talonCommand } = context;
+
+    this.talonCommand = talonCommand;
+    this.command = command;
+    this.languageId = activeEditor.document.languageId;
+    this.marks = this.extractTargetedDecorations(targets, navigationMap);
+    this.targets = targets;
+    this.context = context;
+  }
+
+  extractPrimitiveTargetKeys(...targets: PrimitiveTarget[]) {
+    const keys: string[] = [];
+    targets.forEach((target) => {
+      if (target.mark.type === "decoratedSymbol") {
+        const { character, symbolColor } = target.mark;
+        keys.push(NavigationMap.getKey(symbolColor, character));
+      }
+    });
+    return keys;
+  }
+
+  extractTargetKeys(target: Target): string[] {
+    switch (target.type) {
+      case "primitive":
+        return this.extractPrimitiveTargetKeys(target);
+
+      case "list":
+        return target.elements.map(this.extractTargetKeys, this).flat();
+
+      case "range":
+        return this.extractPrimitiveTargetKeys(target.start, target.end);
+
+      default:
+        return [];
+    }
+  }
+
+  extractTargetedDecorations(targets: Target[], navigationMap: NavigationMap) {
+    if (!navigationMap) {
+      return {};
+    }
+
+    const decorationRanges = navigationMap.serializeRanges();
+    const targetedDecorations: DecorationRanges = {};
+    const targetKeys = targets.map(this.extractTargetKeys, this).flat();
+    targetKeys.forEach((key) => {
+      targetedDecorations[key] = decorationRanges[key];
+    });
+    return targetedDecorations;
+  }
+
+  isThatMarkTargeted() {
```

Comment:
This won't work if a "that" mark is in a compound target, right?  I think you need to descend like you do for finding decorated marks